src/tools/rbd_mirror/group_replayer/Replayer.cc

```
@@ -0,0 +1,1254 @@
+        break;
+      }
+      remove_mirror_peer_uuid(local_snap->id);
+    }
+    locker.unlock();
+    // prune all the image snaps of the group snap locally
+    if (prune_all_image_snapshots(&(*local_snap))) {
+      continue;
+    }
+    dout(10) << "all image snaps are pruned, finally unlinking group snap: "
+             << local_snap->id << dendl;
+    locker.lock();
+    r = librbd::cls_client::group_snap_remove(&m_local_io_ctx,
+        librbd::util::group_header_name(m_local_group_id), local_snap->id);
+    if (r < 0) {
+      derr << "failed to remove group snapshot : "
+           << local_snap->id << " : " << cpp_strerror(r) << dendl;
+    }
+    m_local_group_snaps.erase(local_snap);
+    locker.unlock();
+  }
+}
+
+template <typename I>
+void Replayer<I>::create_regular_snapshot(
+    const std::string &remote_group_snap_name,
+    const std::string &remote_group_snap_id,
+    Context *on_finish) {
+  dout(10) << remote_group_snap_id << dendl;
+  librados::ObjectWriteOperation op;
+  cls::rbd::GroupSnapshot group_snap{
+    remote_group_snap_id, // keeping it same as remote group snap id
+    cls::rbd::GroupSnapshotNamespaceUser{},
+      remote_group_snap_name,
+      cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+  m_local_group_snaps.push_back(group_snap);
+
+  librbd::cls_client::group_snap_set(&op, group_snap);
+  auto comp = create_rados_callback(
+      new LambdaContext([this, on_finish](int r) {
+        handle_create_regular_snapshot(r, on_finish);
+      }));
+  int r = m_local_io_ctx.aio_operate(
+      librbd::util::group_header_name(m_local_group_id), comp, &op);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void Replayer<I>::handle_create_regular_snapshot(
+    int r, Context *on_finish) {
+  dout(10) << "r=" << r << dendl;
+
+  if (r < 0) {
+    derr << "error creating local non-primary group snapshot: "
+         << cpp_strerror(r) << dendl;
+  }
+  on_finish->complete(0);
```

Should this be `on_finish->complete(r)` ? Why ignore error during `group_snap_set`?